BUG: Correctly localize naive datetime strings with Series and datetimetztype (#17415)

mroeschke opened the pull request:
- [x] closes #17415
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

mroeschke committed:
BUG: Correctly localize naive ts strings with Series and datetimetz dtype (#17415)

Add whatsnew

Fix lint error

Rebase and move whatsnew note

Move whatsnew and add tests

pep8speaks commented:
Hello @mroeschke! Thanks for submitting the PR.

 - In the file [`pandas/core/dtypes/cast.py`](https://github.com/pandas-dev/pandas/blob/b2a3b0d2379be8231420ce49ab03ac7fd68ebd57/pandas/core/dtypes/cast.py), following are the PEP8 issues :

> [Line 974:72](https://github.com/pandas-dev/pandas/blob/b2a3b0d2379be8231420ce49ab03ac7fd68ebd57/pandas/core/dtypes/cast.py#L974): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -965,12 +965,16 @@ def maybe_cast_to_datetime(value, dtype, errors='raise'):
-                                     )
+                            is_dt_string = is_string_dtype(value)
+                            value = to_datetime(value, errors=errors)
+                            if is_dt_string:
```
hmm adding this incremental logic is error prone. I think this should simply be converted higher up. can you see if this can be simplied that way.

mroeschke replied:
This logic is pretty high up the stack already.

The Series constructor calls [`_sanitize_array`](https://github.com/pandas-dev/pandas/blob/83436af8ae1ccad49b7ceac7471c060d823d10ab/pandas/core/series.py#L257) which then calls [`maybe_cast_to_datetime`](https://github.com/pandas-dev/pandas/blob/83436af8ae1ccad49b7ceac7471c060d823d10ab/pandas/core/series.py#L2985) (which I am editing here) fairly early. The line I am editing is the first instance where arrays of numbers (and strings) are converted with `to_datetime`

jreback replied:
try passing utc=True to to_datetime

this logic is too specific here

jreback replied:
actually it might be that

to_datetime(...., tz=dtype.tz) might be the correct idiom here (another PR implements this)

mroeschke replied:
Yeah #17413 will make this easier.

However I think there will need to be logic to determine if the incoming data is naive or already localized to UTC. The code comments that were already here note that (numeric) data is already UTC while string data (what I am trying to fix here) may be naive.

jreback replied:
my point is solving this here is wrong 

this is a bug in to_datetime itself

jreback replied:
passing utc=True to to_datetime will make this work (then just converting to the dtype)

jreback replied:
can you try passing ``utc=True`` here

mroeschke replied:
`utc=True` does not work here because the timestamp string should be localized to the specified dtype tz directly intead of UTC in between.

```
# with utc=True
In [5]: ts_naive = '2013-01-01 00:00:00'
   ...:  result = Series([ts_naive], dtype='datetime64[ns, CET]')
   ...:  expected = Series([pd.Timestamp(ts_naive, tz='CET')])

In [6]: result
Out[6]: 
0   2013-01-01 01:00:00+01:00
dtype: datetime64[ns, CET]

In [7]: expected
Out[7]: 
0   2013-01-01 00:00:00+01:00
dtype: datetime64[ns, CET]
```

jreback commented:
pls rebase and move note to 0.21.1

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -488,6 +488,7 @@ Conversion
 - Bug in :func:`Series.fillna` returns frame when ``inplace=True`` and ``value`` is dict (:issue:`16156`)
 - Bug in :attr:`Timestamp.weekday_name` returning a UTC-based weekday name when localized to a timezone (:issue:`17354`)
 - Bug in ``Timestamp.replace`` when replacing ``tzinfo`` around DST changes (:issue:`15683`)
```
move to 0.21.1

jreback commented:
let me have a look

jreback commented:
can you rebase

jreback commented on doc/source/whatsnew/v0.21.1.txt:
```diff
@@ -65,6 +65,7 @@ Conversion
 - Bug in :meth:`IntervalIndex.copy` when copying and ``IntervalIndex`` with non-default ``closed`` (:issue:`18339`)
 - Bug in :func:`DataFrame.to_dict` where columns of datetime that are tz-aware were not converted to required arrays when used with ``orient='records'``, raising``TypeError` (:issue:`18372`)
 - Bug in :class:`DateTimeIndex` and :meth:`date_range` where mismatching tz-aware ``start`` and ``end`` timezones would not raise an err if ``end.tzinfo`` is None (:issue:`18431`)
+- Bug in localization of a naive, datetime string in a ``Series`` constructor with a ``datetime[ns, timezone]`` dtype (:issue:`174151`)
```
move to 0.22

jreback commented on pandas/tests/series/test_constructors.py:
```diff
@@ -574,6 +574,12 @@ def test_constructor_with_datetime_tz(self):
 
+        # GH 17415: With naive string
+        ts_naive = '2013-01-01 00:00:00'
+        result = Series([ts_naive], dtype='datetime64[ns, CET]')
```
make this a separate test

parametrize over adding a NaT, nan, None here as well

mroeschke replied:
```
In [3]: Series([pd.NaT], dtype='datetime64[ns, CET]') # same for np.nan and None
Out[3]: 
0   NaT
dtype: datetime64[ns, CET]

In [4]: Series(pd.Timestamp(pd.NaT, tz='CET')) # same for np.nan and None
Out[4]: 
0   NaT
dtype: datetime64[ns]
```

So I figure both of these should be equivalent and `In[4]` should be the correct behavior?

jreback reviewed (changes requested):
ok I guess this is ok, until we have ``tz=`` in ``pd.to_datetime`` this must be done here. some comments.

jreback commented:
if you can change this to 0.22 for the whatsnew and rebase can have a look

mroeschke committed:
Add missing arg

mroeschke commented:
Rebased and moved the whatsnew to v0.22.

Still curious how to handle the following situation:

```
In [3]: Series([pd.NaT], dtype='datetime64[ns, CET]') # same for np.nan and None
Out[3]: 
0   NaT
dtype: datetime64[ns, CET]

In [4]: Series(pd.Timestamp(pd.NaT, tz='CET')) # same for np.nan and None
Out[4]: 
0   NaT
dtype: datetime64[ns]
```
`Out[4]` should be the correct behavior for both I assume.

jreback commented:
@mroeschke a way you can handle is like this

```
In [1]: from pandas.api.types import DatetimeTZDtype

In [2]: DatetimeTZDtype('datetime64[ns, CET]')
Out[2]: datetime64[ns, CET]

In [3]: DatetimeTZDtype('CET')
ValueError: DatetimeTZDtype constructor must have a tz supplied
```

so in a try/except if you succeed you have a valid dtype, otherwise you can pass it directly to the constructor.

mroeschke committed:
fix failing test and rebase

mroeschke commented:
@jreback  After some digging, the reason why `Out[3]` occurs above is that this happens:

```
In [3]: pd.to_datetime([pd.NaT]).tz_localize('CET')
Out[3]: DatetimeIndex(['NaT'], dtype='datetime64[ns, CET]', freq=None)
```

Curious to get your opinion if this should raise instead.

jreback commented:
```
In [1]: pd.Timestamp('NaT')
Out[1]: NaT

In [2]: pd.Timestamp('NaT').tz_localize('CET')
Out[2]: NaT
```

```
In [4]: pd.to_datetime(['nat']).tz_localize('CET')
Out[4]: DatetimeIndex(['NaT'], dtype='datetime64[ns, CET]', freq=None)

In [5]: pd.to_datetime(['nat', '20170101']).tz_localize('CET')
Out[5]: DatetimeIndex(['NaT', '2017-01-01 00:00:00+01:00'], dtype='datetime64[ns, CET]', freq=None)

In [6]: pd.to_datetime(['nat']).tz_localize('CET')[0]
Out[6]: NaT
```
@mroeschke this is pretty consistent. NaT can be part of a tz-aware Index/Series. It doesn't have a tz itself though.

mroeschke commented:
Makes sense, thanks for confirming.

mroeschke committed:
Correct to  a more appropriate test

mroeschke commented:
I realized I was concerned about a non-issue since:

```
In [4]: pd.to_datetime(['nat']).tz_localize('CET')
Out[4]: DatetimeIndex(['NaT'], dtype='datetime64[ns, CET]', freq=None)
```

makes sense then

```
In [3]: Series([pd.NaT], dtype='datetime64[ns, CET]') # same for np.nan and None
Out[3]: 
0   NaT
dtype: datetime64[ns, CET]
```

makes sense as well.

All green.

mroeschke committed:
Rebase and move whatsnew to v0.23

jreback commented on doc/source/whatsnew/v0.23.0.txt:
```diff
@@ -380,7 +380,7 @@ Conversion
 - Bug in :class:`DatetimeIndex` where the repr was not showing high-precision time values at the end of a day (e.g., 23:59:59.999999999) (:issue:`19030`)
 - Bug where dividing a scalar timedelta-like object with :class:`TimedeltaIndex` performed the reciprocal operation (:issue:`19125`)
--
+- Bug in localization of a naive, datetime string in a ``Series`` constructor with a ``datetime[ns, timezone]`` dtype (:issue:`174151`)
```
``datetime64[ns, tz]``

jreback commented on pandas/core/dtypes/cast.py:
```diff
@@ -1003,12 +1003,18 @@ def maybe_cast_to_datetime(value, dtype, errors='raise'):
-                                     .tz_localize('UTC')
-                                     .tz_convert(dtype.tz)
-                                     )
+                            # This block can be simplified once PR #17413 is
```
add to the comment what that PR is doing

jreback replied:
add an xref (with this PR) to that issue as well

jreback replied:
change to #13712 here (the issue and not the PR which was closed)

jreback reviewed (changes requested):
some comments. otherwise lgtm. ping on green.

mroeschke committed:
address comments

jreback commented:
thanks!
